packages/patterns/integration/instantiate-recipe.test.ts

```diff
@@ -0,0 +1,119 @@
+import { env } from "@commontools/integration";
+import { sleep } from "@commontools/utils/sleep";
+import {
+  registerCharm,
+  ShellIntegration,
+} from "@commontools/integration/shell-utils";
+import { beforeAll, describe, it } from "@std/testing/bdd";
+import { join } from "@std/path";
+import { assert, assertEquals } from "@std/assert";
+import {
+  clickShadowElement,
+  typeInShadowInput,
+} from "./shadow-dom-utils.ts";
+
+const { API_URL, FRONTEND_URL } = env;
+
+// Shadow DOM paths for ct-message-input in CommonTools shell
+const CT_MESSAGE_INPUT_PATH = [
+  "x-root-view",
+  "x-app-view",
+  "x-body-view",
+  "x-charm-view",
+  "common-charm",
+  "ct-render",
+  "ct-message-input",
+  "ct-input",
+  "input",
+];
+
+const CT_MESSAGE_INPUT_BUTTON_PATH = [
+  "x-root-view",
+  "x-app-view",
+  "x-body-view",
+  "x-charm-view",
+  "common-charm",
+  "ct-render",
+  "ct-message-input",
+  "ct-button",
```

Comment:
Most of this is iterating through the shell components. I see the need for wanting to specify/disambiguate a specific node that cannot be represented as a traditional selector, but Astral pierce selectors could query e.g. the single `ct-render` element, query a `ct-message-input` within, or with a unique attribute, `ct-button[main-button=""]` could be queried from the root without specifying the full path -- are there elements which cannot be selected by the pierce strategy in astral?

Reply:
I couldn't ever get the `{strategy: 'piece'}` to actually traverse this path no matter what I tried, it returns `undefined` in almost every case so I gave up and wrote my own solution. 

> `ct-button[main-button=""]`

I'll try this, I don't really like it but I don't like my current solve either so... lesser of two evils?